lib/quda_arpack_interface.cpp

```diff
@@ -1,384 +1,541 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+#include <quda_internal.h>
 #include <quda_arpack_interface.h>
-#include <thrust/sort.h>
+#include <eigensolve_quda.h>
+#include <color_spinor_field.h>
+#include <blas_quda.h>
+#include <util_quda.h>
+#include <sys/time.h>
 
-#if (defined (QMP_COMMS) || defined (MPI_COMMS))
-#include <mpi.h>
-#include <mpi_comm_handle.h>
-#endif
+// ARPACK INTERAFCE ROUTINES
+//--------------------------------------------------------------------------
 
+namespace quda
+{
 
-//using namespace quda ;
+#ifdef ARPACK_LIB
 
-  struct SortEvals{
-    double _val;
-    int    _idx;
+  void arpackErrorHelpNAUPD();
+  void arpackErrorHelpNEUPD();
 
-    static bool small_values;
+#if (defined(QMP_COMMS) || defined(MPI_COMMS))
+#include <mpi.h>
+#endif
 
-    SortEvals(double val, int idx) : _val(val), _idx(idx) {};
+  void arpack_solve(void *h_evecs, void *h_evals, const DiracMatrix &mat, QudaEigParam *eig_param,
+                    ColorSpinorParam *cpuParam)
+  {
 
-    static bool SelectSmall (SortEvals v1, SortEvals v2) { return (v1._val < v2._val);}
-    static bool SelectLarge (SortEvals v1, SortEvals v2) { return (v1._val > v2._val);}
+    // ARPACK logfile name
+    char *arpack_logfile = eig_param->arpack_logfile;
+    if (getVerbosity() >= QUDA_SUMMARIZE) {
+      printfQuda("**** START ARPACK SOLUTION ****\n");
+      printfQuda("Output directed to %s\n", arpack_logfile);
+    }
 
-  };
+    // Create Eigensolver object for member function use
+    TimeProfile profile("Dummy");
+    EigenSolver *eig_solver = EigenSolver::create(eig_param, mat, profile);
+
+    // Construct parameters and memory allocation
+    //---------------------------------------------------------------------------------
+    double time_ar = 0.0; // time in ARPACK
+    double time_mv = 0.0; // time in QUDA mat vec + data transfer
+    double time_ev = 0.0; // time in computing Eigenvectors
+
+    // MPI objects
+    int *fcomm_ = nullptr;
+#if (defined(QMP_COMMS) || defined(MPI_COMMS))
+    MPI_Fint mpi_comm_fort = MPI_Comm_c2f(MPI_COMM_WORLD);
```

Comment:
This needs to be `MPI_COMM_HANDLE` w/regards to the multiple communicator support.

Reply:
In lieu of adding this comment anywhere else (without me checking if it actually does appear anywhere else), just run a `grep -r "MPI_COMM_WORLD"` and make sure it doesn't show up anywhere except for setting the default value of `MPI_COMM_HANDLE` :)

Reply:
Fixed in next push